packages/wrangler/src/dev/dev.tsx

```diff
@@ -295,35 +370,67 @@ function DevSession(props: DevSessionProps) {
 				inspector: {
 					port: props.inspectorPort,
 				},
-				urlOverrides: {
+				origin: {
 					secure: props.localProtocol === "https",
 					hostname: props.localUpstream,
 				},
 				liveReload: props.liveReload,
 			},
-		}),
-		[
-			props.name,
-			props.initialIp,
-			props.initialPort,
-			props.localProtocol,
-			props.httpsKeyPath,
-			props.httpsCertPath,
-			props.localUpstream,
-			props.inspectorPort,
-			props.liveReload,
-		]
-	);
+		} satisfies StartDevWorkerOptions;
+	}, [
+		props.name,
+		props.compatibilityDate,
+		props.compatibilityFlags,
+		props.bindings,
+		props.routes,
+		props.env,
+		props.legacyEnv,
+		props.sendMetrics,
+		props.usageModel,
+		props.isWorkersSite,
+		props.assetPaths,
+		accountIdDeferred,
+		props.local,
+		props.initialIp,
+		props.initialPort,
+		props.localProtocol,
+		props.httpsKeyPath,
+		props.httpsCertPath,
+		props.localUpstream,
+		props.inspectorPort,
+		props.liveReload,
+		props.queueConsumers,
+		props.crons,
+	]);
 
 	const onBundleStart = useCallback(() => {
 		devEnv.proxy.onBundleStart({
 			type: "bundleStart",
 			config: startDevWorkerOptions,
 		});
 	}, [devEnv, startDevWorkerOptions]);
+	const onBundleComplete = useCallback(
+		(bundle: EsbuildBundle) => {
+			if (props.experimentalDevenvRuntime) {
+				devEnv.runtimes.forEach((runtime) =>
```

Comment:
Is there a chance of a race condition here? I'm not completely clear yet, but I'm wondering if something like this could happen:

1. User sets remote flag.
2. Bundler starts with remote runtime `onBundleComplete`.
3. User sets local flag.
4. Bundler starts with local runtime `onBundleComplete`.
5. Remote runtime `onBundleComplete` fires.

Perhaps we can add an E2E test to confirm the intended behaviour here to avoid regressions.

Reply:
I agree we should have a e2e test for this common/likely scenario.

I've made an effort to keep as many methods synchronous (vs async) to avoid race conditons. Which is the case here. The async work inside those event handlers will be the only parts that need to consider races and we do that with equality checks on bundle/config with the latest event (currently accomplished with incrementing ids)

Reply:
@andyjessop I'll add an e2e test to verify this, but the behaviour should be:
1. User sets remote flag.
2. Bundler triggers remote runtime `onBundleComplete` (and local runtime, which is ignored)
3. User sets local flag.
4. Bundler triggers local runtime `onBundleStart`  (and remote runtime, cancelling any in-progress operations)
5. Bundler triggers local runtime `onBundleComplete`  (and remote runtime, which is ignored, cancelling any in-)
6. Remote runtime finishes (if cancellation failed), but doesn't fire `reloadComplete` because the current bundle has changed.

Reply:
d676c6d6847007f4798089b66d33d2e6c754a471